gateway/handlers/update_handler.go

```diff
@@ -1,16 +1,130 @@
 package handlers
 
 import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
 	"net/http"
+	"time"
 
 	"github.com/alexellis/faas/gateway/metrics"
+	"github.com/alexellis/faas/gateway/requests"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/swarm"
 	"github.com/docker/docker/client"
 )
 
-// MakeUpdateFunctionHandler request to update an existing function with new configuration such as image, parameters etc.
+const updateConfigROLLBACK = 2 // From https://github.com/docker/swarmkit/blob/master/api/types.proto#L357
+
+// MakeUpdateFunctionHandler request to update an existing function with new configuration such as image, envvars etc.
 func MakeUpdateFunctionHandler(metricsOptions metrics.MetricOptions, c *client.Client, maxRestarts uint64) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
+		ctx := context.Background()
+
 		defer r.Body.Close()
-		w.WriteHeader(http.StatusNotImplemented)
+		body, _ := ioutil.ReadAll(r.Body)
+
+		request := requests.CreateFunctionRequest{}
+		err := json.Unmarshal(body, &request)
+		if err != nil {
+			log.Println("Error parsing request:", err)
+			w.WriteHeader(http.StatusBadRequest)
+			w.Write([]byte(err.Error()))
+			return
+		}
+
+		serviceInspectopts := types.ServiceInspectOptions{
+			InsertDefaults: true,
+		}
+
+		service, _, err := c.ServiceInspectWithRaw(ctx, request.Service, serviceInspectopts)
+		if err != nil {
+			log.Println("Error inspecting service", err)
+			w.WriteHeader(http.StatusNotFound)
+			w.Write([]byte(err.Error()))
+			return
+		}
+
+		updateSpec(&request, &service.Spec, maxRestarts)
+
+		updateOpts := types.ServiceUpdateOptions{}
+		updateOpts.RegistryAuthFrom = types.RegistryAuthFromSpec
+
+		if len(request.RegistryAuth) > 0 {
+			auth, err := BuildEncodedAuthConfig(request.RegistryAuth, request.Image)
+			if err != nil {
+				log.Println("Error building registry auth configuration:", err)
+				w.WriteHeader(http.StatusBadRequest)
+				w.Write([]byte("Invalid registry auth"))
+				return
+			}
+			updateOpts.EncodedRegistryAuth = auth
+		}
+
+		response, err := c.ServiceUpdate(ctx, service.ID, service.Version, service.Spec, updateOpts)
+		if err != nil {
+			log.Println("Error updating service:", err)
+			w.WriteHeader(http.StatusBadRequest)
+			w.Write([]byte("Update error: " + err.Error()))
+			return
+		}
+		log.Println(response.Warnings)
+	}
+}
+
+func updateSpec(request *requests.CreateFunctionRequest, spec *swarm.ServiceSpec, maxRestarts uint64) {
+
+	linuxOnlyConstraints := []string{"node.platform.os == linux"}
```

Comment:
extract this variable since it's common outside of scope of function.

Reply:
will do